### PR TITLE
add parameterized circuits

### LINKED ISSUE #2055: to support data section in qobj to run the same circuits multiple times with different input data
### What is the expected behavior?
I expect that with the same circuits, I can run it multiple times with different input data. In this case, I don't need to recompile the circuits multiple times.  Every time when I want to run on new data, I just need to update the data section without updating other parts.
For example, in this qasm below, I had one operation "u3(data:input_magic1,0,pi) q1[0]" and a data section with foreach. So if it will execute all circuits for every different input_magic1, it will be very great.

OPENQASM 2.0;\ninclude \"qelib1.inc\";\nqreg q0[2];\nqreg q1[2];\ncreg c0[2];\ncreg c1[2];\nmeasure q1[1] -> c1[1];\nu3(data:input_magic1,0,pi) q1[0];\nmeasure q1[0] -> c1[0];\nu3(pi,0,pi) q0[1];\nmeasure q0[1] -> c0[1];\nu2(localinputvar1,0) q0[0];\nmeasure q0[0] -> c0[0];
section data: foreach (input_magic1: pi; input_magic1: 0.1; input_magic1: 0.2, input_magic1: 0.3)

### PR DESCRIPTION
**Pull request by ewinston**
### Summary
This PR adds the ability to define variable place holders for gate parameters. 

### Details and comments

To create a circuit with a variable parameter use `sympy.Symbol`;
```
import sympy
from qiskit import QuantumCircuit, QuantumRegister
theta = sympy.Symbol('theta')
qr = QuantumRegister(1)
qc = QuantumCircuit(qr)
qc.rx(theta, qr)
```
When an instruction with a Symbol is appended to the circuit, the circuit tracks the parameter in a table which can be accessed with
```
qc.variable_table
```
To see a list of variables in the variable table you can do,
```
qc.variables
```
To set all occurrences of the variable to a constant in the circuit,
```
qc.variable_table[theta] = 0.0
```
The assignment operation does not currently remove the variable from the variable table so the same circuit object can be used to set multiple values, perhaps each of which is converted to qobj for passing to a backend. 

Parameter expressions are also supported, 
```
theta = sympy.Symbol('theta')
phi = sympy.Symbol('phi')
qc.rx(-2 * theta, qr)
qc.u2(theta + phi, theta, qr)
```
To preserve expressions when creating parameterized circuits it may be preferable to evaluate the expression on copies of the circuit. To facilitate this a new method is added to QuantumCircuit so one may do,
```
phi_range = numpy.linspace(0, numpy.pi, 20)
circuits = []
for value in phi_range:
    circuits.append(qc.assign_variable({phi: value, theta: numpy.pi / 4}))
```

Closes #2055

### CODE REVIEW

**Comment by dongreenberg** [comment]
What's the added benefit of using sympy's symbol instead of a string?

**Comment by ewinston** [comment]
It's useful to have a dedicated type for a variable parameter. Using sympy also allows defining functions of symbols which could be useful as demonstrated by the examples in the description.

**Comment by dongreenberg** [comment]
How does this handle a case of a few hundred parameters in a circuit? Often we're parameterizing in the context of variational forms where we don't care to name individually. In that case I'd rather create a placeholder and then fill everything by passing an array of parameters. Can this handle that case?

**Comment by dongreenberg** [comment]
I see. Cool, being able to define functions of parameters is very rad.

**Comment by ewinston** [comment]
Maybe you already answered this but the symbol is the placeholder and the `assign_variable` method of a circuit evaluates/assigns all values, returning the new circuit.

File: qiskit/circuit/variabletable.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        return self._table[key]
+
+    def __setitem__(self, key, value):
+        """set the value of variable
+
+        Args:
+            key (Object): the variable to set
+            value (Number or dict): numeric constant or dictionary of
+                (Symbol:value) pairs
+        """
```
Comment (kdk) [inline qiskit/circuit/variabletable.py]: Worth commenting here that the two operations bounds to `=` are very different, depending on whether or not `key` exists in the table. Might be worth splitting these into two separate methods, one to register a new variable->gate mapping, and another to assign values to a circuit variable.

File: qiskit/circuit/instruction.py [hunk 2]
```diff
@@ -119,6 +96,51 @@ def _define(self):
...
+                self._params.extend(sub_instr.params)  # recursive call
+            return self._params
+        else:
+            return self._params
+
+    @params.setter
+    def params(self, parameters):
+        self._params = []
+        for single_param in parameters:
+            # example: u2(pi/2, sin(pi/4))
```
[the remark this answers is not in the document]
Comment (kdk) [reply qiskit/circuit/instruction.py]: The only difference seems to be whether or not parameters from the definition are automatically pulled in before assignments from `parameters`. It's not clear to me how parameter ownership should work for composite gates. When do we want to expose the params of sub-gates?

**Comment by dongreenberg** [comment]
Sorry - I still don't quite follow. If I have a variational form circuit that I want to load in with 100 or so values (the original motivation for us lobbying for this feature), is there a simple way for me to construct my circuit with "default" placeholders (because I don't really need the names) and then passing an array of values into assign_values? Or, more broadly, is there an interface definition for this feature somewhere?

**Comment by kdk** [comment]
Hi @dongreenberg , I agree that this is difficult with the current API (for example, below), as `assign_values` requires the circuit user to directly manage the symbols, and map them correctly to their values. So in the context of a circuit factory, the user would have to be passed back a set of symbols along with their circuit (or fetch them by inspecting the variable table directly).

An interface where only the circuit object is required to keep record of which symbols it references, and allows a user to assign them either by name or by index should be possible. We would have to decide how the circuit would determine what parameter names or ordering should be exposed to its user (maybe by more explicit parameter declaration), but I don't think this has been defined elsewhere. Suggestions welcome!

```
s = sympy.symbols('p0:99')
qc = qk.QuantumCircuit(qr, cr)

for (theta, phi, lambda) in zip(s[::3], s[1::3], s[2::3]):
  qc.u3(theta, phi, lambda, qr[0])

param_values = [...]
qc.assign_values({ s: v for (s,v) in zip(s, param_values) })
```

**Comment by dongreenberg** [comment]
1) I see that the variable_table is underlied by a dict, which for python 3.6+ is ordered. An easy way to provide 1/2 of the interface convenience I'm mentioning is to allow assign_values to accept a list and two fill modes, "overwrite" and "fill" - If "overwrite", simply fill the values of the variable_table dict in order with the values of my list. If "fill", fill the unassigned values in the variable_table dict with the values in my array. Other options could be to only provide the "fill" option, and force the user to call some "clear_variables()" function if they'd like to overwrite, or provide only an overwrite=True/False option (I think I personally prefer this last option).
2) I don't know anything about sympy, and don't want to learn about a large library to do this simple core activity in Terra. I think most users would prefer to be able to simply pass a string or nothing at all in order to build a gate with a placeholder. If I pass a string, is probably means I'm not going to do any fancy symbolic calculation with it, and I'd prefer for you to build your sympy things in the background. If I pass nothing, I'd like you to autoname it because I'm going to just fill it with an array later. If I add two circuits with some autoname variables together, I assume that the concat function will make sure some are renamed so they don't collide.

With both of those provisions, here's what your code above would look like:
```
qc = qk.QuantumCircuit(qr, cr)
for i in range(var_form_depth):
  qc.u3(0.0, " ", " ", qr[0])
  # in principle this should just be qc.u3(theta=0.0, qr[0]) because the args are no longer required.
param_values = [...]
qc.assign_values(param_values, overwrite=True)
```

Note that you wouldn't add a bunch of u3s to a single gate like this in the wild, you'd be adding them over all your qubits (however that's done today, below is pseudocode) interleaved with CX's:

```
singles = qc.u3(0.0, " ", " ", qr)
doubles = qc.cx(qr[1:]) + qc.cx(qr[:-1])
varcircuit = singles
for i in range(depth):
    varcircuit += doubles + singles
return varcircuit

# Later

my_params = [...]
if not len(my_params) == varcircuit.num_variables():
   raise Error('I screwed up')
varcircuit.assign_variables(my_params)
backend.run(varcircuit)
```

**Comment by kdk** [comment]
:+1: Awesome, thanks for the feedback.

For 1, I agree, I think that's a level of convenience we'd like to have. I'd like something more explicit than exposing the order of the internal dict, which would order parameters according how they were added to the circuit, which to me, seems non-obvious and weakens the circuit as a meaningful level of abstraction. Similarly for fill, the target of parameter assignments would depend on what variables are currently bound, which could lead to very hard to track down or silent bugs. I agree though for cases where its non-ambiguous.

For 2, also very much agree that to the degree we can simplify the parameter interface, to the point that sympy can be pushed behind the scenes or removed, we should. I have similar reservations about e.g. auto-naming variables, and I don't think we're yet at the point where we can forego declaration of parameters entirely, but agree that the process could be simplified for both the circuit creator and the circuit consumer. I would rather start with an explicit interface, and later provide shorthand where convenient and unambiguous. 

What about the following?

```
# Circuit definition
params = sympy.symbols('p0:100')               # Could be a qiskit Parameter class, or similar
varcircuit = qk.QuantumCircuit(qr, cr, params) # Shape of params determines valid signature of assign_variables

singles = varcircuit.u3
doubles = varcircuit.cx(qr[1:]) + varcircuit.cx(qr[:-1])  # Tangential, but this isn't syntax we have currently, would you expect n-1 CX(i, i+1)'s from this ?

varcircuit += singles.assign_variables(0.0, params[0], params[1])
for i in range(1,depth):
    varcircuit += doubles + singles.assign_variables(0.0, params[2*i], params[2*i +1])

# Later 

another_qc = qk.QuantumCircuit(qr, cr)
another_qc =+ my_qc.assign_variables(my_params)
```

Ordered parameters seem to make generalized circuit composition (as in your example) harder, because, without some canonical ordering, it relies on the user of the circuit knowing how the circuit was composed. I'm not sure how to preserve that while keeping a predictable parameter assignment signature.

File: qiskit/circuit/variabletable.py (hunk 1 above)
Comment (ewinston) [reply qiskit/circuit/variabletable.py]: sounds good

**Comment by dongreenberg** [comment]
I see your point about potential bug prickliness. I think at the end of the day having to work through sympy is the biggest block for me - if instructions could be added with strings to denote variables I think the model would be very clean and easy. Adding the layer of sympy adds the behavior of a library I don't understand between me and creating a placeholder parameter. I'd be happy to go collect real user input if you think that'd be informative here. 

I don't completely understand your code sample. Maybe I'll come find you this week to talk it through.

Also @ewinston - is there a reason we went with the name "variable" and not "parameter"? In Aqua and in most of the literature I've seen (e.g. the SVM paper and the hardware efficient ansatz chemistry paper), parameter is the common name for these placeholders.

(@kdk ya, the cx code wouldn't work, I was just shingling them over the qubits, but this isn't a thing in Terra.)

**Pull request merged**